File: elixir/proverb/lib/proverb.ex
defmodule Proverb do
  @doc """
  Generate a proverb from a list of strings.
  """
  @spec recite(strings :: [String.t()]) :: String.t()
  def recite([]), do: ""

  def recite(strings) do
    [first | _] = strings

    for [this, that] <- Enum.chunk_every(strings, 2, 1, :discard),
        reduce: ""
    do
      acc -> acc <> "For want of a #{this} the #{that} was lost.\n"
    end
    <> "And all for the want of a #{first}.\n"
  end
end

File: lua/proverb/proverb.lua
local Proverb = {}

function Proverb.recite(strings)
  local first = function(i)
    return string.format("For want of a %s the %s was lost.\n", strings[i-1], strings[i])
  end
  local last = function()
    return string.format("And all for the want of a %s.\n", strings[1])
  end
  
  local lines = {}
  for i = 2, #strings do table.insert(lines, first(i)) end
  if #strings > 0 then table.insert(lines, last()) end
  
  return table.concat(lines)
end

return Proverb
